connect/proxy/listener_test.go

```diff
@@ -131,8 +135,10 @@ func TestPublicListener(t *testing.T) {
 	// Check active conn is tracked in gauges
 	assertCurrentGaugeValue(t, sink, "consul.proxy.test.inbound.conns;dst=db", 1)
 
-	// Close listener to ensure all conns are closed and have reported their
-	// metrics
+	// Short sleep to allow for cw.written to increment for tx and rx before calling reportStats
+	time.Sleep(time.Millisecond)
```

Comment:
This is cool, but I'm still puzzled how atomic increments appear to update asynchronously... Those atomic increments occur at same time `Write` was called on the packet. I _assume_ `Write` must have returned by the time the test passes since we assert the packet was received.

But we call `reportStats` in a `defer` in the handler loop which will only be executed when one of these cases triggers:

https://github.com/hashicorp/consul/blob/3e6313bebbf087129c35fb1f93b4f69ec48fd91d/connect/proxy/listener.go#L206-L216

`<-connStop` implies that the `Conn` is closed and both actual connections with it so `Write` _must_ have completed

`<-stopChan` is more subtle, it will return when `Listener.Stop` closes the chan. Which means the `Conn` is still running.

We then execute defers in reverse order so `reportStats` gets called before `conn.Close()`.

So the only thing that I can see that could possibly explain flakiness is that somehow, the test code gets the bytes delivered over the network and returns from the `testEchoConn` function, _before_ the `Write` call on the `countWriter` finishes executing.

That is possible but seems incredible that it would happen as much as it does but my mental model for the internals of go network stack may well be off.

Can we just super quickly validate the theory though. All you'd need to do would be to remove the sleep here and then modify: https://github.com/hashicorp/consul/blob/3e6313bebbf087129c35fb1f93b4f69ec48fd91d/connect/proxy/conn.go#L91-L96

To increment the counter, _before_ the `Write`. That assumes the processor can't re-order those instructions though too which I'm not sure if it's valid.

If that change appears to fix it then I'll be a bit happier knowing where the race is. It's not a real "fix" though since we can't count sent bytes until they are really sent since the Write might fail etc.

I think either way, solving it with this sleep is the right approach for now, it's not a race that is ever of importance except in this test. But understanding exactly where the ordering assumptions are violated would be nice. We could update the comment to be more specific too.

If that doesn't fix it then I'm baffled but don't want to spend more time on this now!

Reply:
I removed the sleep and flipped the order inside of Write and now the test passes consistently. 

So yes it does seem to be that testEchoConn returns before Write in these conditions with limited CPU.